Fix prefix match for root path (`/`) to match all paths

nak3 opened the pull request:
7d4a1366f5ce1ca8eef67ccd05975f4c02850e8b fixed the prefix match behavior.

While the fix is correct for `/<something>`, the root path (`/`) does not work as expected.
The root path `/` with prefix match needs to match "all paths" as per doc[1][2].

This patch changed `/` to match all paths.

[1] https://kubernetes.io/docs/concepts/services-networking/ingress/#examples
[2] https://gateway-api.sigs.k8s.io/references/spec/

[X] Networking

Pull Request Attributes

Please check any characteristics that apply to this pull request. 

[X] Does not have any [user-facing](https://github.com/istio/istio/tree/master/releasenotes#when-to-add-release-notes) changes. This may include CLI changes, API changes, behavior changes, performance improvements, etc.

/cc @Monkeyanator @howardjohn

istio-testing commented:
@nak3: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
release-notes_istio | bcfe8c182523cdd7663a6661bec9c107e3b6e437 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/34378/release-notes_istio/1420216426316173312) | `/test release-notes_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

nak3 commented:
/test release-notes_istio
Set `release-notes-none` label as this change roll backs the behavior for root path as the current release (before 7d4a136 ).

istio-testing commented:
@nak3: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integ-pilot-k8s-tests_istio | bcfe8c182523cdd7663a6661bec9c107e3b6e437 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/34378/integ-pilot-k8s-tests_istio/1420216419831779328) | `/test integ-pilot-k8s-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

Monkeyanator commented on pilot/pkg/config/kube/gateway/conversion.go:
```diff
@@ -868,6 +868,11 @@ func createURIMatch(match k8s.HTTPRouteMatch) (*istio.StringMatch, *ConfigError)
 	case k8s.PathMatchPrefix, k8s.PathMatchImplementationSpecific:
+		if dest == "/" {
+			return &istio.StringMatch{
+				MatchType: &istio.StringMatch_Regex{Regex: regexp.QuoteMeta(dest)},
```
should this be `&istio.StringMatch_Prefix`?

nak3 replied:
Thank you. Updated.

Monkeyanator reviewed (commented):
Thanks for reporting @nak3. Could you give an example of a case that doesn't work for `Prefix: /`?  Also adding an integration test for the broken case [here](https://github.com/istio/istio/blob/42b9a0342ef9ea4c38e39821332b41af1c7a96cf/tests/integration/pilot/ingress_test.go#L313) would be helpful.

From experimenting just now I actually found that paths consisting of just backslashes (e.g. `//`, `///`) don't match prefix `/` which is probably an issue since I believe that's technically valid.

nak3 commented:
> Could you give an example of a case that doesn't work for Prefix: /?

Sure, I am using Gateway API and, for example, if I changed HTTPRoute in
https://istio.io/latest/docs/tasks/traffic-management/ingress/gateway-api/ to:

```
apiVersion: networking.x-k8s.io/v1alpha1
kind: HTTPRoute
metadata:
  name: http
  namespace: default
  labels:
    selected: "yes"
spec:
  gateways:
    allow: All
  hostnames: ["httpbin.example.com"]
  rules:
  - matches:
    - path:
        type: Prefix
        value: /           ## Replace with root path
    filters:
    - type: RequestHeaderModifier
      requestHeaderModifier:
        add:
          my-added-header: added-value
    forwardTo:
    - serviceName: httpbin
      port: 8000
```

Then, the curl gets `404`.

```
$ curl -s -I -H "Host: httpbin.example.com" 172.20.0.3:30419/get
HTTP/1.1 404 Not Found
date: Thu, 29 Jul 2021 00:08:53 GMT
server: istio-envoy
transfer-encoding: chunked
```

The gateway api doc says [1], so the curl for `/get` should work with prefix `/`, correct?

[1] https://gateway-api.sigs.k8s.io/references/spec/
> If no matches are specified, the default is a prefix path match on “/”, which has the effect of matching every HTTP request.

istio-testing commented:
@nak3: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
unit-tests_istio | e720e975b714fa5a515e47bab9febc7828028524 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/34378/unit-tests_istio/1420536385047105536) | `/test unit-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@nak3: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
unit-tests_istio | e720e975b714fa5a515e47bab9febc7828028524 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/34378/unit-tests_istio/1420536385047105536) | `/test unit-tests_istio`
integ-pilot-k8s-tests_istio | e720e975b714fa5a515e47bab9febc7828028524 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/34378/integ-pilot-k8s-tests_istio/1420536385126797312) | `/test integ-pilot-k8s-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@nak3: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
unit-tests_istio | e720e975b714fa5a515e47bab9febc7828028524 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/34378/unit-tests_istio/1420536385047105536) | `/test unit-tests_istio`
integ-pilot-k8s-tests_istio | e720e975b714fa5a515e47bab9febc7828028524 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/34378/integ-pilot-k8s-tests_istio/1420536385126797312) | `/test integ-pilot-k8s-tests_istio`
integ-cni-k8s-tests_istio | e720e975b714fa5a515e47bab9febc7828028524 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/34378/integ-cni-k8s-tests_istio/1420536385399427072) | `/test integ-cni-k8s-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@nak3: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
unit-tests_istio | e720e975b714fa5a515e47bab9febc7828028524 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/34378/unit-tests_istio/1420536385047105536) | `/test unit-tests_istio`
integ-pilot-k8s-tests_istio | e720e975b714fa5a515e47bab9febc7828028524 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/34378/integ-pilot-k8s-tests_istio/1420536385126797312) | `/test integ-pilot-k8s-tests_istio`
integ-cni-k8s-tests_istio | e720e975b714fa5a515e47bab9febc7828028524 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/34378/integ-cni-k8s-tests_istio/1420536385399427072) | `/test integ-cni-k8s-tests_istio`
integ-pilot-multicluster-tests_istio | e720e975b714fa5a515e47bab9febc7828028524 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/34378/integ-pilot-multicluster-tests_istio/1420536386691272704) | `/test integ-pilot-multicluster-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@nak3: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integ-pilot-k8s-tests_istio | e720e975b714fa5a515e47bab9febc7828028524 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/34378/integ-pilot-k8s-tests_istio/1420536385126797312) | `/test integ-pilot-k8s-tests_istio`
integ-cni-k8s-tests_istio | e720e975b714fa5a515e47bab9febc7828028524 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/34378/integ-cni-k8s-tests_istio/1420536385399427072) | `/test integ-cni-k8s-tests_istio`
integ-pilot-multicluster-tests_istio | e720e975b714fa5a515e47bab9febc7828028524 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/34378/integ-pilot-multicluster-tests_istio/1420536386691272704) | `/test integ-pilot-multicluster-tests_istio`
unit-tests_istio | 82619278e5603b0355af2d8cab1f76feb345dc10 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/34378/unit-tests_istio/1420553715516968960) | `/test unit-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

Monkeyanator commented:
Ah I think the real problem here is we don't trim the `/` suffix in:

https://github.com/istio/istio/blob/master/pilot/pkg/config/kube/gateway/conversion.go#L872

Maybe we should have `regexp.QuoteMeta(strings.TrimSuffix(dest, "/"))` or similar rather than having special handling for `Prefix: /`? From local testing that seems to get things working.

nak3 commented:
Oh! Yes, that works on my env. Thank you! Updated.

nak3 commented:
@howardjohn Could you please take a look?

istio-testing commented:
In response to a cherrypick label: #34378 failed to apply on top of branch "release-1.10":
```
Applying: Fix prefix match for root path (`/`).
Using index info to reconstruct a base tree...
M	pilot/pkg/config/kube/gateway/conversion.go
M	pilot/pkg/config/kube/gateway/testdata/http.yaml
M	pilot/pkg/config/kube/gateway/testdata/http.yaml.golden
M	tests/integration/pilot/ingress_test.go
Falling back to patching base and 3-way merge...
Auto-merging tests/integration/pilot/ingress_test.go
Auto-merging pilot/pkg/config/kube/gateway/testdata/http.yaml.golden
Auto-merging pilot/pkg/config/kube/gateway/testdata/http.yaml
Auto-merging pilot/pkg/config/kube/gateway/conversion.go
CONFLICT (content): Merge conflict in pilot/pkg/config/kube/gateway/conversion.go
error: Failed to merge in the changes.
hint: Use 'git am --show-current-patch=diff' to see the failed patch
Patch failed at 0001 Fix prefix match for root path (`/`).
When you have resolved this problem, run "git am --continue".
If you prefer to skip this patch, run "git am --skip" instead.
To restore the original branch and stop patching, run "git am --abort".

```

istio-testing commented:
In response to a cherrypick label: new issue created for failed cherrypick: #34504

istio-testing commented:
In response to a cherrypick label: new pull request created: #34505
